x/protorev/keeper/hooks_test.go

```diff
@@ -0,0 +1,775 @@
+			s.Require().NoError(err)
+			s.Require().Equal(tc.param.expectedTrades, routes.Trades)
+		})
+	}
+}
+
+// Tests the hook implementation that is called after liquidity providing
+func (s *KeeperTestSuite) TestLiquidityChanging() {
+	type param struct {
+		expectedTrades            []types.Trade
+		executeLiquidityProviding func()
+	}
+
+	tests := []struct {
+		name       string
+		param      param
+		expectPass bool
+	}{
+		{
+			name: "GAMM - Join Swap Exact Amount In",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+				},
+				executeLiquidityProviding: func() {
+					_, err := s.App.GAMMKeeper.JoinSwapExactAmountIn(s.Ctx, s.TestAccs[0], 1, sdk.NewCoins(sdk.NewCoin("akash", sdk.NewInt(100))), sdk.NewInt(1))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "GAMM - Join Swap Share Amount Out",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+				},
+				executeLiquidityProviding: func() {
+					_, err := s.App.GAMMKeeper.JoinSwapShareAmountOut(s.Ctx, s.TestAccs[0], 1, "akash", sdk.NewInt(1000), sdk.NewInt(10000))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "GAMM - Exit Swap Exact Amount Out",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+				},
+				executeLiquidityProviding: func() {
+					_, err := s.App.GAMMKeeper.ExitSwapExactAmountOut(s.Ctx, s.TestAccs[0], 1, sdk.NewCoin("Atom", sdk.NewInt(1)), sdk.NewInt(1002141106353159235))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "GAMM - Exit Swap Share Amount In",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+				},
+				executeLiquidityProviding: func() {
+					_, err := s.App.GAMMKeeper.ExitSwapShareAmountIn(s.Ctx, s.TestAccs[0], 1, "Atom", sdk.NewInt(1000000000000000000), sdk.NewInt(1))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "GAMM - Exit Swap Share Amount In - Low Shares",
+			param: param{
+				expectedTrades: []types.Trade(nil),
+				executeLiquidityProviding: func() {
+					_, err := s.App.GAMMKeeper.ExitSwapShareAmountIn(s.Ctx, s.TestAccs[0], 1, "Atom", sdk.NewInt(1000), sdk.NewInt(0))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			tc.param.executeLiquidityProviding()
+
+			routes, err := s.App.ProtoRevKeeper.GetSwapsToBackrun(s.Ctx)
+			if tc.expectPass {
+				s.Require().NoError(err)
+				s.Require().Equal(tc.param.expectedTrades, routes.Trades)
+			}
+
+			s.App.ProtoRevKeeper.DeleteSwapsToBackrun(s.Ctx)
```

Comment:
nit: what is the reason for manually deleting this? We could also just call `s.Setup() at the start of the test case?

Reply:
It was done originally because the last two tests called the ExitSwapShareAmountIn methods that required the test acc to have the pool shares in their balance, which a previous iteration of the test gave them.

Instead, I have now just given the testAccs the necessarily pool shares balance and just call s.SetupTest(): https://github.com/osmosis-labs/osmosis/pull/5045/commits/460da4a42d4c3cf341890d0537982419dd6aee51